File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/Handlers/StockValidationHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example1.Handlers;

public class StockValidationHandler : BaseOrderHandler
{
    private readonly Dictionary<string, int> _stockInventory;

    public StockValidationHandler()
    {
        _stockInventory = new Dictionary<string, int>
        {
            { "Laptop", 10 },
            { "Mouse", 50 },
            { "Keyboard", 30 },
            { "Monitor", 5 },
            { "Headphones", 0 }
        };
    }

    public override void Handle(OrderRequest request)
    {
        if (!_stockInventory.ContainsKey(request.ProductName))
        {
            request.AddMessage($"Stock validation failed: Product '{request.ProductName}' not found");
            request.IsApproved = false;
            return;
        }

        var availableStock = _stockInventory[request.ProductName];

        if (availableStock >= request.Quantity)
        {
            _stockInventory[request.ProductName] -= request.Quantity;
            request.AddMessage($"Stock validated: {request.Quantity} units available");
            base.Handle(request);
        }
        else
        {
            request.AddMessage($"Insufficient stock: Need {request.Quantity}, available {availableStock}");
            request.IsApproved = false;
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/src/Handlers/StockValidationHandler.ts
import { BaseOrderHandler } from "../BaseOrderHandler";
import { OrderRequest } from "../OrderRequest";

export class StockValidationHandler extends BaseOrderHandler {
	private readonly _stock: Map<string, number> = new Map([
		["Laptop",     10],
		["Mouse",      50],
		["Keyboard",   30],
		["Monitor",     5],
		["Headphones",  0],
	]);

	handle(request: OrderRequest): void {
		const available = this._stock.get(request.productName);
		if (available === undefined) {
			request.addMessage(`Stock validation failed: Product '${request.productName}' not found`);
			request.isApproved = false;
			return;
		}
		if (available >= request.quantity) {
			this._stock.set(request.productName, available - request.quantity);
			request.addMessage(`Stock validated: ${request.quantity} units available`);
			super.handle(request);
		} else {
			request.addMessage(`Insufficient stock: Need ${request.quantity}, available ${available}`);
			request.isApproved = false;
		}
	}
}
